### ISSUE TITLE
VB spec says that Me.M() operates on a COPY of Me in a struct

### ISSUE THREAD

**Issue by gafter** [dotnet/vblang#52]
The current state of the VB spec says that "Me" is a value, not a variable.

Therefore, in a struct type an invocation such as

Me.M()

Is required to operate on a copy of the current object and not modify it.

Worse, the simple invocation

M()

is defined to operate on the method group Me.M, so it too may not modify the current object.

 Migrated from TFS/DevDiv 1040115
 Migrated from https://github.com/dotnet/roslyn/issues/4686

**Comment by gafter** [comment]
@gafter wrote

> I suggest that Me should be specified to be a value in a class, but a variable in a struct.

@AlekseyTs wrote

> It looks like value receiver isn't copied if it is backed by a variable or it is "Me", see bug #1026678.

**Comment by AnthonyDGreen** [comment]
I know you can't assign to Me in a strict. That's an open design question. Any objection to making Me assignable in structs (as it will be classified as a variable)?

**Comment by KlausLoeffelmann** [comment]
Would it make this behave differently (actually break), since not a copy would be passed to Bar?

If it would, please don't change it.

```
Public Structure StructTest

    Public Property IntProp As Integer

    Public Sub Foo()
        'Not allowed, assigned to a value.
        'Me = New StructTest
        Bar(Me)
    End Sub

    Public Sub Bar(ByRef sTest As StructTest)
        sTest = New StructTest With {.IntProp = 42}
    End Sub
End Structure

```

**Comment by AnthonyDGreen** [comment]
Quite possibly. Does that code exist?

**Comment by AnthonyDGreen** [comment]
Hmm, I tried @KlausLoeffelmann's example and agree it's too subtle to break the current behavior. I'll just have to find a way to spec it.

Looking at the code I suspect the special-casing of invocations happens here http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/CodeGen/EmitExpression.vb,1005

```VB.NET
                ElseIf IsVerifierValue(receiverType) Then
 
                    Dim methodContainingType = method.ContainingType
                    If IsVerifierValue(methodContainingType) AndAlso MayUseCallForStructMethod(method) Then
 
                        ' NOTE: this should be either a method which overrides some abstract method or 
                        '       does not override anything (with few exceptions, see MayUseCallForStructMethod); 
                        '       otherwise we should not use direct 'call' and must use constrained call;
 
                        ' calling a method defined in a value type considered a "write" to the target unless the target is "Me"
                        Debug.Assert(receiverType = method.ContainingType)
                        tempOpt = EmitReceiverRef(
                            receiver,
                            isAccessConstrained:=False, addressKind:=If(IsMeReceiver(receiver),
                                                                         AddressKind.ReadOnly,
                                                                         AddressKind.Writeable))
```

If I'm reading this right the emitter assumes (when calling an instance **method**) that in a value type a Me reference is read-only and doesn't emit the temp. @AlekseyTs do you agree with that?

If so, I think we can update the spec to say something like 'Me' is a value except when it's the receiver of an instance method invocation or instance field access. Would that correct the spec, @gafter?

**Comment by AlekseyTs** [comment]
> If I'm reading this right the emitter assumes (when calling an instance method) that in a value type a Me reference is read-only and doesn't emit the temp. @AlekseyTs do you agree with that?

Yes, this matches my understanding. It sounds like updating the spec would address the issue.

**Comment by gafter** [comment]
@AnthonyDGreen Yes, I think that spec change would work.
